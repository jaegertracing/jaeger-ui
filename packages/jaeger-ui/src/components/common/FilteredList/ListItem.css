/*
Copyright (c) 2019 Uber Technologies, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.FilteredList--ListItem {
  align-items: center;
  cursor: pointer;
  display: flex;
  padding: 0.5em 5em 0.5em 0.5em;
  white-space: nowrap;
  min-width: 100%;
}

.FilteredList--ListItem.is-striped {
  background: linear-gradient(to right, rgba(0, 0, 0, 0.04), rgba(0, 0, 0, 0));
}

.FilteredList--ListItem.is-striped:hover,
.FilteredList--ListItem:hover {
  background: linear-gradient(
    to right,
    rgba(135, 232, 222, 0.1),
    rgba(135, 232, 222, 0.2) 70%,
    rgba(135, 232, 222, 0)
  );
  color: var(--tx-color-title);
}

.FilteredList--ListItem.is-selected {
  background: linear-gradient(to right, #87e8de, #87e8deaa 70%, #87e8de00);
  color: var(--tx-color-title);
}

.FilteredList--ListItem.is-striped.is-focused,
.FilteredList--ListItem.is-focused {
  background: linear-gradient(
    to right,
    rgba(135, 232, 222, 0.2),
    rgba(17, 153, 153, 0.2) 70%,
    rgba(17, 153, 153, 0)
  );
  color: var(--tx-color-title);
}

.FilteredList--ListItem mark {
  background: none;
  color: #eb2f96;
  font-weight: 500;
  padding: 0;
}
