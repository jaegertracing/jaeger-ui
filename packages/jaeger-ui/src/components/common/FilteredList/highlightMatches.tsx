// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import memoize from 'lru-memoize';

type TIsHighlightGroupFn = (index: number) => boolean;

type THighlightMatchFn = (match: RegExpMatchArray) => React.ReactNode;

function toHighlights(isHighlightGroup: TIsHighlightGroupFn, match: RegExpMatchArray) {
  const texts = match
    .map((tx: string, i: number) => {
      if (i === 0 || !tx) {
        // the first group is the full match, not a capturing group
        return null;
      }
      if (isHighlightGroup(i)) {
        return <mark key={`${tx + i}`}>{tx}</mark>;
      }
      return tx;
    })
    .filter(Boolean);
  return <span>{texts}</span>;
}

// In the matches from the regular expressions created in `MatchHighlighter`,
// text that should be highlighted is interleaved with text that shouldn't be
// highlighted. E.g. alternating capturing groups should be highlighted,
// regardless of how many capturing groups there are in the match. Therefore, we
// need to establish if the first capturing group (i.e. match[1]) should be
// highlighted or if it should start at match[2]. Then, we alternate, from there.
// This results in either highlighting even groups (starting at match[2]) or odd
// groups (starting at match[1]). However, we don't actually need to highlight
// odd groups, but just the first group (match[1]), so we can simplify that test.
const highlightEvenGroups = toHighlights.bind(null, (index: number) => index % 2 === 0);
const highlightFirstGroup = toHighlights.bind(null, (index: number) => index === 1);

const ANY_LAZY = '(.*?)';
const ANY_GREEDY = '(.*)';

function wordBreak(lowerLetter: string) {
  return `(\\b${lowerLetter}|${lowerLetter.toUpperCase()})`;
}

function letterClass(lowerLetter: string) {
  return `[${lowerLetter}${lowerLetter.toUpperCase()}]`;
}

/**
 * Create regular expressions used for highlighting matches between the query
 * and the text. This is derived from the match-sorter ranking system description:
 * https://github.com/kentcdodds/match-sorter#this-solution
 *
 * Several regular expressions are created from the filterText. These roughly
 * align with the ranking system described in the link. They are used to divide
 * input text into emphasized and non-emphasized sections.
 */
class MatchHighlighter {
  public static make(query: string) {
    return new MatchHighlighter(query);
  }

  private readonly matchers: [RegExp, THighlightMatchFn][];

  constructor(readonly query: string) {
    // istanbul ignore next
    if (!query) {
      this.matchers = [];
      return;
    }
    const tx = query.toLowerCase();
    const initialCap = `${tx[0].toUpperCase()}${tx.slice(1)}`;
    const letters = tx.split('');
    const startsWith = `^(${tx})`;
    const wordStartsWith = `(\\b${letters.map(letterClass).join('')}|${initialCap})`;
    const acronymLetters = letters.map(letter => wordBreak(letter)).join(ANY_LAZY);
    const constains = `(${tx})`;
    const anyLetters = letters.map(letter => `(${letter})`).join(ANY_LAZY);

    // given the query "eg", the following regular expressions will be created:
    this.matchers = [
      // case-insensitive match at start of text, would match "ego" but not "lego"
      // /^(eg)(.*)/i
      [new RegExp(`${startsWith}${ANY_GREEDY}`, 'i'), highlightFirstGroup],

      // case-insensitive match at start of word, with boundaries in camelCase counting as words
      // would match "theEgg" and "the-egg"
      // /(.*?)(\b[eE][gG]|Eg)(.*)/
      [new RegExp(`${ANY_LAZY}${wordStartsWith}${ANY_GREEDY}`), highlightEvenGroups],

      // match an acronym, e.g. "easy-going" and "easyGoing" would both match "eg"
      // /(.*?)(\be|E)(.*?)(\bg|G)(.*)/
      [new RegExp(`${ANY_LAZY}${acronymLetters}${ANY_GREEDY}`), highlightEvenGroups],

      // case insensitive contains
      // /(.*?)(eg)(.*)/i
      [new RegExp(`${ANY_LAZY}${constains}${ANY_GREEDY}`, 'i'), highlightEvenGroups],

      // contains each letter, in sequence
      // /(.*?)(e)(.*?)(g)(.*)/i
      [new RegExp(`${ANY_LAZY}${anyLetters}${ANY_GREEDY}`, 'i'), highlightEvenGroups],
    ];
  }

  highlightMatches(text: string) {
    if (!text) {
      return text;
    }
    for (let i = 0; i < this.matchers.length; i++) {
      const [matcher, highlightMatch] = this.matchers[i];
      const match = text.match(matcher);
      if (match) {
        return highlightMatch(match);
      }
    }
    return text;
  }
}

const getHighlighter = memoize(30)(MatchHighlighter.make);

function highlightMatchesImpl(query: string, text: string) {
  if (!query) {
    return text;
  }
  return getHighlighter(query).highlightMatches(text);
}

export default memoize(200)(highlightMatchesImpl);
