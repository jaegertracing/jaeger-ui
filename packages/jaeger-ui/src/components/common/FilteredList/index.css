/*
Copyright (c) 2019 Uber Technologies, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.FilteredList--list {
  background: #fafafa;
}

.FilteredList--filterWrapper {
  align-items: center;
  display: flex;
  box-shadow: 0 0 3px 1px rgba(0, 0, 0, 0.3);
  position: relative;
  z-index: 10;
}

.FilteredList--filterIcon {
  font-size: 1.3em;
  margin: 0 0.25em 0 0.65em;
  position: absolute;
}

.FilteredList--filterInput {
  border: none;
  flex: 1;
  height: auto;
  padding: 0.5em 0.3em 0.5em 3em;
}

.FilteredList--filterInput:focus {
  box-shadow: inset 0px 0px 3px 1px #6aa;
  outline: 1px solid #199;
  outline-offset: -1px;
}
