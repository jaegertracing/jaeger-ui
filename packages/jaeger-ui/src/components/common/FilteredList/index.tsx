// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import _debounce from 'lodash/debounce';
import matchSorter from 'match-sorter';
import IoIosSearch from 'react-icons/lib/io/ios-search';
import { FixedSizeList as VList, ListOnItemsRenderedProps, ListOnScrollProps } from 'react-window';
import { Key as EKey } from 'ts-key-enum';

import ListItem from './ListItem';

import './index.css';

type TProps = {
  cancel: () => void;
  options: string[];
  value: string | null;
  setValue: (value: string) => void;
};

type TState = {
  filterText: string;
  visibleStartIndex: number;
  visibleStopIndex: number;
  focusedIndex: number | null;
};

export default class FilteredList extends React.PureComponent<TProps, TState> {
  inputRef: React.RefObject<HTMLInputElement> = React.createRef();
  vlistRef: React.RefObject<VList> = React.createRef();
  state: TState = {
    filterText: '',
    focusedIndex: null,
    visibleStartIndex: 0,
    visibleStopIndex: 0,
  };

  componentDidUpdate() {
    this.focusInput();
  }

  focusInput = () => {
    const { current } = this.inputRef;
    if (current) {
      current.focus();
    }
  };

  private getFilteredOptions = () => {
    const { options } = this.props;
    const { filterText } = this.state;
    return filterText ? matchSorter(options, filterText) : options;
  };

  private setValue = (value: string) => {
    this.props.setValue(value);
    this.setState({ filterText: '', focusedIndex: null });
  };

  private onKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
    const { focusedIndex: stFocused } = this.state;
    switch (true) {
      case event.key === EKey.Escape: {
        const { cancel } = this.props;
        this.setState({ filterText: '', focusedIndex: null });
        cancel();
        return;
      }
      case event.key === EKey.ArrowUp:
      case event.key === EKey.ArrowDown: {
        const { visibleStartIndex, visibleStopIndex } = this.state;
        let focusedIndex: number | void;
        if (stFocused == null) {
          focusedIndex = event.key === EKey.ArrowDown ? visibleStartIndex : visibleStopIndex;
          this.setState({ focusedIndex });
        } else {
          const offset = event.key === EKey.ArrowDown ? 1 : -1;
          const filteredOptions = this.getFilteredOptions();
          const i = stFocused + offset;
          focusedIndex = i > -1 ? i % filteredOptions.length : filteredOptions.length + i;
          this.setState({ focusedIndex });
        }
        const listInstance = this.vlistRef.current;
        if (listInstance && (focusedIndex < visibleStartIndex + 1 || focusedIndex > visibleStopIndex - 1)) {
          listInstance.scrollToItem(focusedIndex);
        }
        return;
      }
      case event.key === EKey.Enter: {
        if (stFocused == null) {
          return;
        }
        const filteredOptions = this.getFilteredOptions();
        const value = filteredOptions[stFocused];
        this.setValue(value);
        break;
      }
      default: // no-op
    }
  };

  private onListScrolled = _debounce((scrollInfo: ListOnScrollProps) => {
    if (!scrollInfo.scrollUpdateWasRequested) {
      this.setState({ focusedIndex: null });
    }
  }, 80);

  private onListItemsRendered = _debounce((viewInfo: ListOnItemsRenderedProps) => {
    const { visibleStartIndex, visibleStopIndex } = viewInfo;
    this.setState({ visibleStartIndex, visibleStopIndex });
  }, 80);

  private onFilterChanged = (event: React.ChangeEvent<HTMLInputElement>) => {
    this.setState({ filterText: event.target.value, focusedIndex: null });
  };

  render() {
    const { value } = this.props;
    const { filterText, focusedIndex } = this.state;
    const filteredOptions = this.getFilteredOptions();
    const data = {
      focusedIndex,
      highlightQuery: filterText,
      options: filteredOptions,
      selectedValue: value,
      setValue: this.setValue,
    };
    return (
      <div>
        <label className="FilteredList--filterWrapper">
          <IoIosSearch className="FilteredList--filterIcon" />
          <input
            className="FilteredList--filterInput"
            placeholder="Filter..."
            onChange={this.onFilterChanged}
            onKeyDown={this.onKeyDown}
            ref={this.inputRef}
            type="text"
            value={filterText}
          />
        </label>
        <VList
          key={filterText}
          className="FilteredList--list u-simple-scrollbars"
          height={375}
          itemCount={filteredOptions.length}
          itemData={data}
          itemSize={35}
          width={650}
          overscanCount={25}
          onItemsRendered={this.onListItemsRendered}
          onScroll={this.onListScrolled}
          ref={this.vlistRef}
        >
          {ListItem}
        </VList>
      </div>
    );
  }
}
