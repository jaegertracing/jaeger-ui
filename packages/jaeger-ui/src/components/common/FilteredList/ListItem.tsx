// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import cx from 'classnames';
import { ListChildComponentProps } from 'react-window';

import highlightMatches from './highlightMatches';

import './ListItem.css';

interface IListItemProps extends ListChildComponentProps {
  data: {
    focusedIndex: number | null;
    highlightQuery: string;
    options: string[];
    selectedValue: string | null;
    setValue: (value: string) => void;
  };
}

export default class ListItem extends React.PureComponent<IListItemProps> {
  onClicked = () => {
    const { data, index } = this.props;
    const { options, setValue } = data;
    const value = options[index];
    setValue(value);
  };

  render() {
    const { data, index, style: styleOrig } = this.props;
    // omit the width from the style so the panel can scroll horizontally
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const { width: _, ...style } = styleOrig;
    const { focusedIndex, highlightQuery, options, selectedValue } = data;
    const cls = cx('FilteredList--ListItem', {
      'is-focused': index === focusedIndex,
      'is-selected': options[index] === selectedValue,
      'is-striped': index % 2,
    });
    return (
      <div
        className={cls}
        style={style}
        onClick={this.onClicked}
        role="switch"
        aria-checked={index === focusedIndex ? 'true' : 'false'}
      >
        {highlightMatches(highlightQuery, options[index])}
      </div>
    );
  }
}
