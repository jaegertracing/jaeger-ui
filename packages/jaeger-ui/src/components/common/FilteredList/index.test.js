// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React from 'react';
import { shallow } from 'enzyme';
import { FixedSizeList as VList } from 'react-window';
import { Key as EKey } from 'ts-key-enum';

import FilteredList from './index';

describe('<FilteredList>', () => {
  const words = ['and', 'apples', 'are'];
  const numbers = ['0', '1', '2'];

  let props;
  let wrapper;

  const getData = () => wrapper.find(VList).prop('itemData');

  const keyDown = key => wrapper.find('input').simulate('keydown', { key });

  beforeEach(() => {
    props = {
      cancel: jest.fn(),
      options: words.concat(numbers),
      value: null,
      setValue: jest.fn(),
    };
    wrapper = shallow(<FilteredList {...props} />);
  });

  it('renders without exploding', () => {
    expect(wrapper.exists()).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });

  it('puts the focus on the input on update', () => {
    const fn = jest.fn();
    wrapper.instance().inputRef = {
      current: {
        focus: fn,
      },
    };
    wrapper.setProps({ value: 'anything' });
    expect(fn.mock.calls.length).toBe(1);
  });

  it('filters options based on the current input text', () => {
    expect(getData().options).toEqual(props.options);
    wrapper.find('input').simulate('change', { target: { value: 'a' } });
    expect(getData().options).toEqual(words);
  });

  it('setting the value clears the filter and focus index', () => {
    const n = -99;
    const s = 'a';
    wrapper.setState({ filterText: s, focusedIndex: n });
    let data = getData();
    expect(data.focusedIndex).toBe(n);
    expect(data.options).toEqual(words);

    data.setValue('anything');
    data = getData();
    expect(data.options.length).toBe(props.options.length);
    expect(data.focusedIndex).toBe(null);
  });

  describe('up / down arrow keys', () => {
    let indices;

    beforeAll(jest.useFakeTimers);

    beforeEach(() => {
      indices = {
        visibleStartIndex: 1,
        visibleStopIndex: props.options.length - 1,
      };
      wrapper.instance().onListItemsRendered(indices);
      jest.runAllTimers();
    });

    afterAll(jest.useRealTimers);

    it('down arrow sets the focus index to the first visible item when focusIndex == null', () => {
      keyDown(EKey.ArrowDown);
      expect(wrapper.state('focusedIndex')).toBe(indices.visibleStartIndex);
    });

    it('up arrow sets the focus index to the last visible item when focusIndex == null', () => {
      keyDown(EKey.ArrowUp);
      expect(wrapper.state('focusedIndex')).toBe(indices.visibleStopIndex);
    });

    it('shift the focus index to the next element', () => {
      keyDown(EKey.ArrowUp);
      expect(wrapper.state('focusedIndex')).toBe(indices.visibleStopIndex);
      keyDown(EKey.ArrowUp);
      expect(wrapper.state('focusedIndex')).toBe(indices.visibleStopIndex - 1);
    });

    it('cause the view to scroll if necessary', () => {
      const fn = jest.fn();
      keyDown(EKey.ArrowDown);
      expect(wrapper.state('focusedIndex')).toBe(indices.visibleStartIndex);
      wrapper.instance().vlistRef = {
        current: {
          scrollToItem: fn,
        },
      };
      keyDown(EKey.ArrowUp);
      expect(wrapper.state('focusedIndex')).toBe(indices.visibleStartIndex - 1);
      expect(fn.mock.calls).toEqual([[indices.visibleStartIndex - 1]]);
    });
  });

  it('escape triggers cancel', () => {
    expect(props.cancel.mock.calls.length).toBe(0);
    keyDown(EKey.Escape);
    expect(props.cancel.mock.calls.length).toBe(1);
  });

  it('enter selects the current focus index', () => {
    const focusedIndex = 0;
    expect(props.setValue.mock.calls.length).toBe(0);
    wrapper.setState({ focusedIndex });
    keyDown(EKey.Enter);
    expect(props.setValue.mock.calls).toEqual([[props.options[focusedIndex]]]);
  });

  it('enter is ignored when an item is not focused', () => {
    expect(props.setValue.mock.calls.length).toBe(0);
    keyDown(EKey.Enter);
    expect(props.setValue.mock.calls.length).toBe(0);
  });

  it('scrolling unsets the focus index', () => {
    jest.useFakeTimers();
    wrapper.setState({ focusedIndex: 0 });
    wrapper.instance().onListScrolled({ scrollUpdateWasRequested: false });
    jest.runAllTimers();
    expect(wrapper.state('focusedIndex')).toBe(null);
  });
});
