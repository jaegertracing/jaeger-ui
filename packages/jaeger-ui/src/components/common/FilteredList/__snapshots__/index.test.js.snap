// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FilteredList> renders without exploding 1`] = `
<div>
  <label
    className="FilteredList--filterWrapper"
  >
    <IoIosSearch
      className="FilteredList--filterIcon"
    />
    <input
      className="FilteredList--filterInput"
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="Filter..."
      type="text"
      value=""
    />
  </label>
  <List
    className="FilteredList--list u-simple-scrollbars"
    direction="ltr"
    height={375}
    itemCount={6}
    itemData={
      Object {
        "focusedIndex": null,
        "highlightQuery": "",
        "options": Array [
          "and",
          "apples",
          "are",
          "0",
          "1",
          "2",
        ],
        "selectedValue": null,
        "setValue": [Function],
      }
    }
    itemSize={35}
    key=""
    layout="vertical"
    onItemsRendered={[Function]}
    onScroll={[Function]}
    overscanCount={25}
    useIsScrolling={false}
    width={650}
  >
    <Component />
  </List>
</div>
`;
